File: android/core/network/src/main/kotlin/xyz/klinik/network/AnalyticsApi.kt
package xyz.klinik.network

import java.math.BigDecimal
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * The clinic dashboard (spec M11, T6.4).
 *
 * The types here exist mostly to stop one mistake: rendering "not enough to
 * say" as a number. A share, a conversion rate and an occupancy rate are all
 * nullable, and each is null for a reason the reader has to be told — too few
 * cases, or no working hours configured. Showing 0% instead would be a
 * confident claim the data does not support.
 */

/** A proportion that may not exist yet. */
@JvmInline
value class Proportion(val value: Double?) {
    val isKnown: Boolean get() = value != null

    /** The percentage, or null for the caller to replace with a reason. */
    val percent: Int? get() = value?.let { Math.round(it * 100).toInt() }
}

@Serializable
data class NamedCount(
    /** The clinic's own spelling. */
    val label: String,
    val count: Int = 0,
    /** Null when there are too few cases to state a proportion. */
    val share: Double? = null,
) {
    val proportion: Proportion get() = Proportion(share)
}

@Serializable
data class MonthlyCount(
    /** `2026-03`. */
    val month: String,
    val count: Int = 0,
)

@Serializable
data class ProcedureReport(
    val from: String,
    val to: String,
    val total: Int = 0,
    /** Every month in range, empty ones included. */
    val byMonth: List<MonthlyCount> = emptyList(),
    val byProcedure: List<NamedCount> = emptyList(),
) {
    val busiestMonth: MonthlyCount? get() = byMonth.maxByOrNull { it.count }
}

@Serializable
data class GeographyReport(
    val from: String,
    val to: String,
    val total: Int = 0,
    val byCountry: List<NamedCount> = emptyList(),
    val byCity: List<NamedCount> = emptyList(),
    /** Patients with no city recorded, so the shares add up. */
    val cityUnknown: Int = 0,
)

@Serializable
data class ChannelRow(
    val label: String,
    val key: String,
    val patients: Int = 0,
    val converted: Int = 0,
    val conversionRate: Double? = null,
    /** Absent when the viewer may not see money. */
    val revenue: Totals? = null,
) {
    val conversion: Proportion get() = Proportion(conversionRate)
}

@Serializable
data class ChannelReport(
    val from: String,
    val to: String,
    val total: Int = 0,
    val channels: List<ChannelRow> = emptyList(),
    /**
     * True when the viewer may not see money.
     *
     * The screen must say this. An empty revenue column otherwise reads as
     * "this channel earned nothing", which is a different and much worse claim
     * than "you are not allowed to see it".
     */
    val revenueWithheld: Boolean = false,
    /** What "converted" means, so two readers cannot mean two things. */
    val conversionDefinition: String = "",
    val minimumForRate: Int = 0,
) {
    val revenueNoticeKey: String? get() = if (revenueWithheld) "analytics_revenue_withheld" else null
}

@Serializable
data class MonthlyNet(
    val month: String,
    val net: String,
    /** False when this month had amounts with no exchange rate. */
    val converted: Boolean = true,
) {
    val value: BigDecimal get() = BigDecimal(net)
}

@Serializable
data class CurrencyAverage(
    val currency: Currency,
    val average: String,
    val count: Int = 0,
) {
    val value: BigDecimal get() = BigDecimal(average)
}

@Serializable
data class RevenueReport(
    val from: String,
    val to: String,
    val currency: Currency,
    val gross: Totals,
    val discount: Totals,
    val net: Totals,
    val cost: Totals,
    val agencyCommission: Totals,
    /** Net less costs and commission. */
    val margin: Totals,
    val byMonth: List<MonthlyNet> = emptyList(),
    /** Exact, per currency. An average blended across currencies is a fiction. */
    val averageByCurrency: List<CurrencyAverage> = emptyList(),
    val recordCount: Int = 0,
    val cancelledExcluded: Int = 0,
    /** Non-zero means the margin is missing something. */
    val unreadableCostLines: Int = 0,
) {
    /** Whether the margin can be shown without a caveat beside it. */
    val marginIsWhole: Boolean get() = unreadableCostLines == 0 && margin.complete

    val caveatKeys: List<String>
        get() = buildList {
            if (unreadableCostLines > 0) add("analytics_cost_lines_unread")
            if (!net.complete) add("finance_totals_incomplete")
        }
}

@Serializable
data class MonthlyOccupancy(
    val month: String,
    val bookedMinutes: Int = 0,
    val availableMinutes: Int = 0,
    /** Null when no working hours are configured — not zero. */
    val rate: Double? = null,
    val appointments: Int = 0,
) {
    val occupancy: Proportion get() = Proportion(rate)
}

@Serializable
data class OccupancyReport(
    val from: String,
    val to: String,
    val byMonth: List<MonthlyOccupancy> = emptyList(),
    /** True when no availability window exists at all. */
    val capacityUnconfigured: Boolean = false,
) {
    /** What to put on the screen instead of a chart nobody can read. */
    val noticeKey: String?
        get() = if (capacityUnconfigured) "analytics_capacity_unconfigured" else null
}

class AnalyticsApi(
    private val client: ApiClient,
    private val json: Json = ApiClient.defaultJson,
) {
    suspend fun procedures(from: String, to: String): ProcedureReport =
        decode(client.send(Endpoint(HttpMethod.GET, "analytics/procedures?${range(from, to)}")))

    suspend fun geography(from: String, to: String): GeographyReport =
        decode(client.send(Endpoint(HttpMethod.GET, "analytics/geography?${range(from, to)}")))

    suspend fun channels(from: String, to: String, currency: Currency = Currency.TRY): ChannelReport =
        decode(
            client.send(
                Endpoint(HttpMethod.GET, "analytics/channels?${range(from, to)}&currency=${currency.name}"),
            ),
        )

    suspend fun revenue(from: String, to: String, currency: Currency = Currency.TRY): RevenueReport =
        decode(
            client.send(
                Endpoint(HttpMethod.GET, "analytics/revenue?${range(from, to)}&currency=${currency.name}"),
            ),
        )

    suspend fun occupancy(from: String, to: String): OccupancyReport =
        decode(client.send(Endpoint(HttpMethod.GET, "analytics/occupancy?${range(from, to)}")))

    private fun range(from: String, to: String): String = "from=$from&to=$to"

    private inline fun <reified T> decode(body: String): T =
        runCatching { json.decodeFromString<T>(body) }
            .getOrElse { throw ApiError.Decoding(it.message ?: "unreadable response") }
}

File: ios/Sources/KlinikAPI/AnalyticsAPI.swift
import Foundation
import KlinikCore

/**
 * The clinic dashboard (spec M11, T6.4).
 *
 * The types here exist mostly to stop one mistake: rendering "not enough to
 * say" as a number. A share, a conversion rate and an occupancy rate are all
 * optional, and every one of them is nil for a reason the reader has to be
 * told — too few cases, or no working hours configured. Printing 0% instead
 * would be a confident claim the data does not support.
 */

/// A proportion that may not exist yet.
public struct Proportion: Sendable, Equatable {
    public let value: Double?

    public init(_ value: Double?) {
        self.value = value
    }

    public var isKnown: Bool { value != nil }

    /**
     * The percentage, or the placeholder — never "0%".
     *
     * `reason` says why there is no number, because "—" on its own invites the
     * reader to assume the worst interpretation.
     */
    public func formatted(reason: String = L10n.string("analytics.tooFew")) -> String {
        guard let value else { return reason }

        return "\(Int((value * 100).rounded()))%"
    }
}

public struct NamedCount: Decodable, Sendable, Equatable, Identifiable {
    /// The clinic's own spelling.
    public let label: String
    public let count: Int
    /// Null when there are too few cases to state a proportion.
    public let share: Double?

    public var id: String { label }
    public var proportion: Proportion { Proportion(share) }
}

public struct MonthlyCount: Decodable, Sendable, Equatable, Identifiable {
    /// `2026-03`.
    public let month: String
    public let count: Int

    public var id: String { month }
}

public struct ProcedureReport: Decodable, Sendable, Equatable {
    public let from: Date
    public let to: Date
    public let total: Int
    /// Every month in range, empty ones included.
    public let byMonth: [MonthlyCount]
    public let byProcedure: [NamedCount]

    public var busiestMonth: MonthlyCount? { byMonth.max(by: { $0.count < $1.count }) }
}

public struct GeographyReport: Decodable, Sendable, Equatable {
    public let from: Date
    public let to: Date
    public let total: Int
    public let byCountry: [NamedCount]
    public let byCity: [NamedCount]
    /// Patients with no city recorded, so the shares add up.
    public let cityUnknown: Int
}

public struct ChannelRow: Decodable, Sendable, Equatable, Identifiable {
    public let label: String
    public let key: String
    public let patients: Int
    public let converted: Int
    public let conversionRate: Double?
    /// Absent when the viewer may not see money — see `ChannelReport.revenueWithheld`.
    public let revenue: Totals?

    public var id: String { key }
    public var conversion: Proportion { Proportion(conversionRate) }
}

public struct ChannelReport: Decodable, Sendable, Equatable {
    public let from: Date
    public let to: Date
    public let total: Int
    public let channels: [ChannelRow]
    /**
     * True when the viewer may not see money.
     *
     * The screen must say this. An empty revenue column otherwise reads as
     * "this channel earned nothing", which is a different and much worse claim
     * than "you are not allowed to see it".
     */
    public let revenueWithheld: Bool
    /// What "converted" means, so two readers cannot mean two things.
    public let conversionDefinition: String
    public let minimumForRate: Int

    public var revenueNotice: String? {
        revenueWithheld ? L10n.string("analytics.revenueWithheld") : nil
    }
}

public struct MonthlyNet: Decodable, Sendable, Equatable, Identifiable {
    public let month: String
    public let net: String
    /// False when this month had amounts with no exchange rate.
    public let converted: Bool

    public var id: String { month }
    public var amount: Amount { Amount(net) }
}

public struct CurrencyAverage: Decodable, Sendable, Equatable, Identifiable {
    public let currency: Currency
    public let average: String
    public let count: Int

    public var id: String { currency.rawValue }
    public var amount: Amount { Amount(average) }
}

public struct RevenueReport: Decodable, Sendable, Equatable {
    public let from: Date
    public let to: Date
    public let currency: Currency
    public let gross: Totals
    public let discount: Totals
    public let net: Totals
    public let cost: Totals
    public let agencyCommission: Totals
    /// Net less costs and commission.
    public let margin: Totals
    public let byMonth: [MonthlyNet]
    /// Exact, per currency. An average blended across currencies is a fiction.
    public let averageByCurrency: [CurrencyAverage]
    public let recordCount: Int
    public let cancelledExcluded: Int
    /// Non-zero means the margin is missing something.
    public let unreadableCostLines: Int

    /// Whether the margin can be shown without a caveat beside it.
    public var marginIsWhole: Bool { unreadableCostLines == 0 && margin.complete }

    public var caveats: [String] {
        var notes: [String] = []

        if unreadableCostLines > 0 { notes.append(L10n.string("analytics.costLinesUnread")) }
        if !net.complete { notes.append(L10n.string("finance.totals.incomplete")) }

        return notes
    }
}

public struct MonthlyOccupancy: Decodable, Sendable, Equatable, Identifiable {
    public let month: String
    public let bookedMinutes: Int
    public let availableMinutes: Int
    /// Null when no working hours are configured — not zero.
    public let rate: Double?
    public let appointments: Int

    public var id: String { month }
    public var occupancy: Proportion { Proportion(rate) }
}

public struct OccupancyReport: Decodable, Sendable, Equatable {
    public let from: Date
    public let to: Date
    public let byMonth: [MonthlyOccupancy]
    /// True when no availability window exists at all.
    public let capacityUnconfigured: Bool

    /// What to put on the screen instead of a chart nobody can read.
    public var notice: String? {
        capacityUnconfigured ? L10n.string("analytics.capacityUnconfigured") : nil
    }
}

public struct AnalyticsAPI: Sendable {
    private let client: APIClient

    public init(client: APIClient) {
        self.client = client
    }

    public func procedures(from: Date, to: Date) async throws -> ProcedureReport {
        try await client.send(
            Endpoint(method: .get, path: "analytics/procedures", query: Self.range(from, to)),
            as: ProcedureReport.self
        )
    }

    public func geography(from: Date, to: Date) async throws -> GeographyReport {
        try await client.send(
            Endpoint(method: .get, path: "analytics/geography", query: Self.range(from, to)),
            as: GeographyReport.self
        )
    }

    public func channels(
        from: Date,
        to: Date,
        currency: Currency = .turkishLira
    ) async throws -> ChannelReport {
        try await client.send(
            Endpoint(
                method: .get,
                path: "analytics/channels",
                query: Self.range(from, to).merging(["currency": currency.rawValue]) { _, new in new }
            ),
            as: ChannelReport.self
        )
    }

    public func revenue(
        from: Date,
        to: Date,
        currency: Currency = .turkishLira
    ) async throws -> RevenueReport {
        try await client.send(
            Endpoint(
                method: .get,
                path: "analytics/revenue",
                query: Self.range(from, to).merging(["currency": currency.rawValue]) { _, new in new }
            ),
            as: RevenueReport.self
        )
    }

    public func occupancy(from: Date, to: Date) async throws -> OccupancyReport {
        try await client.send(
            Endpoint(method: .get, path: "analytics/occupancy", query: Self.range(from, to)),
            as: OccupancyReport.self
        )
    }

    private static func range(_ from: Date, _ to: Date) -> [String: String] {
        let formatter = ISO8601DateFormatter()

        return ["from": formatter.string(from: from), "to": formatter.string(from: to)]
    }
}
